<a href="https://colab.research.google.com/github/padhab34-bit/IML/blob/main/fashion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import time
import tensorflow as tf
from tensorflow.keras import datasets, layers, models
from tensorflow.keras.utils import to_categorical

print("Loading Fashion-MNIST...")
(train_images, train_labels), (test_images, test_labels) = datasets.fashion_mnist.load_data()

# Reshape images to (28, 28, 1) and normalize pixel values to [0, 1]
train_images = train_images.reshape((60000, 28, 28, 1)).astype('float32') / 255
test_images = test_images.reshape((10000, 28, 28, 1)).astype('float32') / 255


train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)


model = models.Sequential([
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model.summary()


model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

print("\nStarting training...")
start_time = time.time()
history = model.fit(train_images, train_labels, epochs=5, batch_size=64, validation_split=0.1)
end_time = time.time()
training_duration = end_time - start_time

print(f"\nTraining completed in {training_duration:.2f} seconds.")

test_loss, test_acc = model.evaluate(test_images, test_labels, verbose=2)
print(f"\nTest accuracy on Fashion-MNIST: {test_acc:.4f}")
test_loss, test_acc = model.evaluate(test_images, test_labels, verbose=2)
print(f"\nTest accuracy on Fashion-MNIST: {test_acc:.4f}")


Loading Fashion-MNIST...


/usr/local/lib/python3.13/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_3 (Conv2D)               │ (None, 26, 26, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 13, 13, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_4 (Conv2D)               │ (None, 11, 11, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_3 (MaxPooling2D)  │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 3, 3, 64)       │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 576)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 64)             │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 93,322 (364.54 KB)

 Trainable params: 93,322 (364.54 KB)

 Non-trainable params: 0 (0.00 B)


Starting training...
Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 49s 57ms/step - accuracy: 0.7986 - loss: 0.5565 - val_accuracy: 0.8623 - val_loss: 0.3782
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 48s 57ms/step - accuracy: 0.8735 - loss: 0.3504 - val_accuracy: 0.8683 - val_loss: 0.3509
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 48s 57ms/step - accuracy: 0.8917 - loss: 0.2989 - val_accuracy: 0.8838 - val_loss: 0.3120
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 82s 57ms/step - accuracy: 0.9049 - loss: 0.2637 - val_accuracy: 0.9007 - val_loss: 0.2761
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 48s 56ms/step - accuracy: 0.9108 - loss: 0.2395 - val_accuracy: 0.9020 - val_loss: 0.2679

Training completed in 274.94 seconds.
313/313 - 3s - 8ms/step - accuracy: 0.8982 - loss: 0.2789

Test accuracy on Fashion-MNIST: 0.8982
313/313 - 3s - 8ms/step - accuracy: 0.8982 - loss: 0.2789

Test accuracy on Fashion-MNIST: 0.8982


### Comparison Discussion:

1. **Performance (Accuracy):**
   MNIST is easier, so a basic CNN can usually reach about 99% accuracy. Fashion-MNIST is harder because clothes have more details and different shapes, so the accuracy is usually around 90% to 92%.

2. **Training Times:**
   Both datasets have the same number of images and the same 28×28 grayscale image size. Because of this, they take almost the same amount of time to train when using the same computer and batch size.

### Experiment 1: Removing the first Dense layer (64 neurons)
Let's build, compile, and train a model where the flattened feature map goes directly to the final 10-unit output layer with softmax activation.

In [3]:
# Re-build Convnet without the Dense(64) layer
model_no_dense = models.Sequential([
    layers.Input(shape=(28, 28, 1)),
    layers.Conv2D(32, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.Flatten(),
    layers.Dense(10, activation='softmax') # Directly to output
])

model_no_dense.compile(optimizer='adam',
                       loss='categorical_crossentropy',
                       metrics=['accuracy'])

print("Training model without the first Dense layer...")
history_no_dense = model_no_dense.fit(train_images, train_labels, epochs=5, batch_size=64, validation_split=0.1)

test_loss_nd, test_acc_nd = model_no_dense.evaluate(test_images, test_labels, verbose=2)
print(f"\nTest accuracy without the first Dense layer: {test_acc_nd:.4f}")

Training model without the first Dense layer...
Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 49s 57ms/step - accuracy: 0.7945 - loss: 0.5661 - val_accuracy: 0.8582 - val_loss: 0.3909
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 81s 55ms/step - accuracy: 0.8709 - loss: 0.3582 - val_accuracy: 0.8805 - val_loss: 0.3246
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 47s 55ms/step - accuracy: 0.8884 - loss: 0.3079 - val_accuracy: 0.8853 - val_loss: 0.3124
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 45s 54ms/step - accuracy: 0.8983 - loss: 0.2793 - val_accuracy: 0.8953 - val_loss: 0.2945
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 45s 54ms/step - accuracy: 0.9059 - loss: 0.2574 - val_accuracy: 0.8908 - val_loss: 0.2925
313/313 - 3s - 10ms/step - accuracy: 0.8892 - loss: 0.3099

Test accuracy without the first Dense layer: 0.8892


In [4]:
# Re-build Convnet with an added Dense(4096) layer
model_large_dense = models.Sequential([
    layers.Input(shape=(28, 28, 1)),
    layers.Conv2D(32, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.Flatten(),
    layers.Dense(4096, activation='relu'), # Added layer
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model_large_dense.compile(optimizer='adam',
                         loss='categorical_crossentropy',
                         metrics=['accuracy'])

print("Training model with an added 4096-unit Dense layer...")
history_large = model_large_dense.fit(train_images, train_labels, epochs=5, batch_size=64, validation_split=0.1)

test_loss_lg, test_acc_lg = model_large_dense.evaluate(test_images, test_labels, verbose=2)
print(f"\nTest accuracy with the added 4096 Dense layer: {test_acc_lg:.4f}")

Training model with an added 4096-unit Dense layer...
Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 94s 109ms/step - accuracy: 0.8134 - loss: 0.5035 - val_accuracy: 0.8695 - val_loss: 0.3565
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 92s 109ms/step - accuracy: 0.8850 - loss: 0.3131 - val_accuracy: 0.8962 - val_loss: 0.2831
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 142s 109ms/step - accuracy: 0.9023 - loss: 0.2639 - val_accuracy: 0.8985 - val_loss: 0.2704
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 142s 109ms/step - accuracy: 0.9126 - loss: 0.2356 - val_accuracy: 0.9023 - val_loss: 0.2616
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 142s 109ms/step - accuracy: 0.9217 - loss: 0.2123 - val_accuracy: 0.9025 - val_loss: 0.2662
313/313 - 4s - 13ms/step - accuracy: 0.8982 - loss: 0.2861

Test accuracy with the added 4096 Dense layer: 0.8982


We performed all the experiments using a Convolutional Neural Network (ConvNet) on the Fashion-MNIST dataset to see how changes in the model structure affected accuracy and training time. The baseline ConvNet achieved approximately 89.81% accuracy and took around 277.05 seconds to train. Fashion-MNIST is more difficult than standard MNIST, which usually reaches about 99% accuracy, because clothing images contain more complex shapes and greater variation within each category. However, the training times are nearly the same because both datasets contain the same number of images and use 28×28 grayscale images. When the 64-unit Dense layer was removed, the accuracy decreased slightly to approximately 88.92% because the model had less capacity to learn higher-level patterns. When a 4096-unit Dense layer was added, the accuracy was approximately 89.82%, which was almost the same as the baseline. This large layer increased the model size from fewer than 100,000 parameters to more than 2.4 million and significantly increased the training time, but it did not provide any meaningful improvement in accuracy.